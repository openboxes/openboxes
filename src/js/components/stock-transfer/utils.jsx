import _ from 'lodash';

const PENDING = 'PENDING';
const CANCELED = 'CANCELED';

function extractItem(item, status) {
  const { destinationBinLocation, destinationZone } = item;

  let destinationBin = {};
  if (status === PENDING) {
    destinationBin = {
      id: null, name: null, zoneId: null, zoneName: null,
    };
  } else {
    destinationBin = {
      id: destinationBinLocation && destinationBinLocation.id ? destinationBinLocation.id : null,
      name: destinationBinLocation && destinationBinLocation.name
        ? destinationBinLocation.name : null,
      zoneId: destinationZone && destinationZone.id ? destinationZone.id : null,
      zoneName: destinationZone && destinationZone.name ? destinationZone.name : null,
    };
  }

  return {
    ...item,
    destinationBinLocation: destinationBin,
    quantity: status === PENDING ? '' : item.quantity,
  };
}

function extractSplitItem(item, splitItem, status) {
  let destinationBin = {};

  if (status === PENDING) {
    destinationBin = {
      id: null, name: null, zoneId: null, zoneName: null,
    };
  } else {
    destinationBin = {
      id: splitItem.destinationBinLocation.id ? splitItem.destinationBinLocation.id : null,
      name: splitItem.destinationBinLocation.name ? splitItem.destinationBinLocation.name : null,
      zoneId: splitItem.destinationZone ? splitItem.destinationZone.id : null,
      zoneName: splitItem.destinationZone ? splitItem.destinationZone.name : null,
    };
  }

  return {
    ...splitItem,
    destinationBinLocation: destinationBin,
    quantity: status === PENDING ? '' : splitItem.quantity,
    quantityOnHand: item.quantityOnHand,
    referenceId: item.id, // set a referenceId from original item
  };
}

export function extractStockTransferItems(stockTransfer) {
  const stockTransferItems = [];
  _.forEach(stockTransfer.stockTransferItems, (item) => {
    stockTransferItems.push(extractItem(item, stockTransfer.status));
    if (item.splitItems.length > 0) {
      _.forEach(item.splitItems, (splitItem) => {
        stockTransferItems.push(extractSplitItem(item, splitItem, stockTransfer.status));
      });
    }
  });
  return stockTransferItems;
}

export function extractNonCanceledItems(stockTransfer) {
  const stockTransferItems = [];
  _.forEach(stockTransfer.stockTransferItems, (item) => {
    if (item.status !== CANCELED) {
      stockTransferItems.push(extractItem(item, stockTransfer.status));
    }
    if (item.splitItems.length > 0) {
      _.forEach(item.splitItems, (splitItem) => {
        stockTransferItems.push(extractSplitItem(item, splitItem, stockTransfer.status));
      });
    }
  });
  return stockTransferItems;
}

export function prepareRequest(stockTransfer, status) {
  const originalItems = _.filter(stockTransfer.stockTransferItems, (item) => !item.referenceId);
  const stockTransferItems = _.map(originalItems, (item) => ({
    ...item,
    splitItems: _.filter(
      stockTransfer.stockTransferItems,
      (splitItem) => splitItem.referenceId === item.id,
    ),
    quantity: item.quantity === '' ? item.quantityOnHand : item.quantity,
  }));

  return { ...stockTransfer, stockTransferItems, status };
}
