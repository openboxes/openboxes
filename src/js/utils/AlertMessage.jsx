import React, { useMemo } from 'react';

import PropTypes from 'prop-types';

const AlertMessage = ({
  show,
  message,
  danger,
  close,
  className,
}) => {
  const getMessageElement = useMemo(() => {
    const alertMessages = typeof message !== 'string' && Array.isArray(message) ? message : [message];
    return alertMessages.map((msg) => (
      <div>
        <i className={`fa ${danger ? 'fa-times-circle' : 'fa-exclamation-circle'} pr-2`} />
        {msg}
      </div>
    ));
  }, [message]);

  if (show) {
    return (
      <div
        data-testid="alert-message"
        className={`${className} alert ${danger ? 'alert-danger' : 'alert-warning'}`}
        style={{
          cursor: close ? 'pointer' : 'default',
          whiteSpace: 'pre-line',
        }}
        role="presentation"
        onClick={() => close()}
      >

        {getMessageElement}
      </div>
    );
  }

  return null;
};

export default AlertMessage;

AlertMessage.propTypes = {
  className: PropTypes.string,
  close: PropTypes.func,
  danger: PropTypes.bool,
  message: PropTypes.string,
  show: PropTypes.bool,
};

AlertMessage.defaultProps = {
  className: '',
  close: () => null,
  danger: false,
  message: '',
  show: false,
};
