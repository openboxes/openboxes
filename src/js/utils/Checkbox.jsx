import React from 'react';

import PropTypes from 'prop-types';

import Translate from 'utils/Translate';

const Checkbox = ({
  value, indeterminate, custom, fieldRef, withLabel, label, defaultMessage, ...props
}) => {
  const onChange = (event) => {
    const { checked } = event.target;

    if (props.onChange) {
      props.onChange(checked);
    }
  };

  if (custom) {
    return (
      <div data-testid="custom-checkbox" className="custom-checkbox">
        <label htmlFor={props.id}>
          <input
            id={props.id}
            type="checkbox"
            ref={fieldRef}
            checked={value}
            {...props}
            onChange={onChange}
          />
        </label>
      </div>
    );
  }

  if (withLabel) {
    return (
      <div data-testid="with-label-checkbox" className="d-flex align-items-center">
        <input
          type="checkbox"
          ref={(elem) => {
            if (elem) {
              // eslint-disable-next-line no-param-reassign
              elem.indeterminate = indeterminate;
            }
            if (fieldRef) {
              fieldRef(elem);
            }
          }}
          checked={value}
          {...props}
          onChange={onChange}
        />
        <label htmlFor={props.id} style={{ margin: '0 0 0 5px' }}>
          <Translate id={label} defaultMessage={defaultMessage || label} />
        </label>
      </div>
    );
  }

  return (
    <input
      data-testid="checkbox"
      type="checkbox"
      ref={(elem) => {
        if (elem) {
          // eslint-disable-next-line no-param-reassign
          elem.indeterminate = indeterminate;
        }
        if (fieldRef) {
          fieldRef(elem);
        }
      }}
      checked={value}
      {...props}
      onChange={onChange}
    />
  );
};

export default Checkbox;

Checkbox.propTypes = {
  onChange: PropTypes.func,
  value: PropTypes.bool,
  indeterminate: PropTypes.bool,
  custom: PropTypes.bool,
  id: PropTypes.string,
  fieldRef: PropTypes.func,
  withLabel: PropTypes.bool,
  label: PropTypes.string,
  defaultMessage: PropTypes.string,
};

Checkbox.defaultProps = {
  onChange: null,
  value: null,
  indeterminate: false,
  custom: false,
  id: '',
  fieldRef: undefined,
  withLabel: false,
  label: '',
  defaultMessage: '',
};
