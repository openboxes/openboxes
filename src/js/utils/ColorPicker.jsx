import React from 'react';

import PropTypes from 'prop-types';
import { SliderPicker } from 'react-color';
import { Tooltip } from 'react-tippy';

import Input from 'utils/Input';

const ColorPicker = ({ value, onChange }) => {
  const handleChange = (color) => {
    const hex = color.hex.split('#')[1];
    onChange(hex);
  };

  const colorHex = value ? `#${value}` : '#ffffff';

  return (
    <Tooltip
      html={(
        <div style={{ width: 350 }}>
          <SliderPicker
            color={colorHex}
            onChangeComplete={handleChange}
          />
        </div>
      )}
      theme="transparent"
      arrow="true"
      trigger="click"
      interactive
      delay="150"
      duration="250"
    >
      <div className="input-group">
        <div className="input-group-prepend">
          <span className="input-group-text" style={{ backgroundColor: colorHex }}>&nbsp;&nbsp;</span>
        </div>
        <Input value={value} onChange={(val) => onChange(val)} />
        <div className="input-group-append">
          <button type="button" className="input-group-text" onClick={() => onChange(null)}>
            <i className="fa fa-close" />
          </button>
        </div>
      </div>
    </Tooltip>
  );
};

export default ColorPicker;

ColorPicker.propTypes = {
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string,
};

ColorPicker.defaultProps = {
  value: null,
};
