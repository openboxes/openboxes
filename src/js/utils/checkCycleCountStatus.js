import CycleCountCandidateStatus from 'consts/cycleCountCandidateStatus';

const COUNTING_STATUSES = [
  CycleCountCandidateStatus.CREATED,
  CycleCountCandidateStatus.REQUESTED,
  CycleCountCandidateStatus.COUNTING,
];

const RESOLVING_STATUSES = [
  CycleCountCandidateStatus.COUNTED,
  CycleCountCandidateStatus.INVESTIGATING,
];

export const isCounting = (status) => COUNTING_STATUSES.includes(status);
export const isResolving = (status) => RESOLVING_STATUSES.includes(status);
