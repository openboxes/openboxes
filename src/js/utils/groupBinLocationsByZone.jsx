import React from 'react';

import _ from 'lodash';

import useTranslate from 'hooks/useTranslate';

const groupBinLocationsByZone = (binLocations) => {
  const translate = useTranslate();
  const groupedByZone = _.groupBy(binLocations, (bin) => bin.zoneId || 'no-zone');

  return Object.entries(groupedByZone)
    .map(([zoneKey, bins]) => {
      const zoneName = bins[0].zoneName || translate('react.cycleCount.noZone', 'No Zone');

      return {
        id: `zone-${zoneKey}`,
        name: zoneName,
        label: <span className="zone-label">{zoneName}</span>,
        isDisabled: true,
        options: bins
          .map((bin) => ({
            id: bin.id,
            name: bin.name,
            label: bin.name,
            value: bin.id,
          }))
          .sort((a, b) => a.name.localeCompare(b.name)),
      };
    })
    .sort((a, b) => {
      if (a.id === 'no-zone') {
        return 1;
      }
      if (b.id === 'no-zone') {
        return -1;
      }
      return a.name.localeCompare(b.name);
    });
};

export default groupBinLocationsByZone;
