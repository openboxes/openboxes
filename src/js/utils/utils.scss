.custom-modal-content {
  min-height: 173px;
  min-width: 600px;
  box-shadow: 0 0 4px 0 #00000040;
  padding: 20px;

  div {
    gap: 15px;
  }

  .custom-modal-title {
    font-size: 20px;
    font-weight: 500;
    color: var(--blue-800);
  }

  .custom-modal-text {
    font-size: 14px;
    font-weight: 400;
    color: var(--gray-600);
  }

  svg {
    color: var(--blue-700);
  }
}

.redirect-button {
  font-size: 18px;
  text-decoration: none;
  color: var(--blue-800);
  font-weight: 500;

  &:hover {
    color: var(--blue-800);
  }

  span {
    cursor: pointer;
  }
}

.create-page-title {
  font-size: 20px;
  font-weight: 600;

  .create-page-tile-main-content {
    color: var(--gray-600);
  }

  .create-page-title-source-code {
    color: var(--gray-500);
  }
}

.product-supplier-buttons-container {
  height: 100px;
}

.text-blue-primary {
  color: var(--blue-primary);
}

.text-blue-400 {
  color: var(--blue-400);
}

.context-menu {
  background-color: white;
  box-shadow: 0 0 4px 1px rgba(0, 0, 0, 0.15);
  border-radius: 4px;

  .dropdown-item {
    color: var(--blue-700) !important;;

    &:hover {
      background-color: var(--blue-300) !important;
    }
  }

  .font-red-ob {
    color: var(--color-red) !important;
  }
}

.react-tiny-popover-container {
  z-index: 999;
}

.badge-container {
  display: flex;
  flex-wrap: wrap;
  align-items: flex-start;
  gap: 2px;

  .badge {
    padding: 4px 10px;
    border-radius: 999px;
    font-size: 12px;
    white-space: normal;
    max-width: 170px;

    &--purple {
      background-color: var(--purple-100);
      color: var(--purple-600);
    }

    &--blue {
      background-color: var(--blue-202);
      color: var(--blue-401);
    }
  }
}
