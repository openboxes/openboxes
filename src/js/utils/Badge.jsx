import React from 'react';

import PropTypes from 'prop-types';

import './utils.scss';

const Badge = ({ label, variant }) => (
  <span className={`badge ${variant}`}>{label}</span>
);

Badge.propTypes = {
  label: PropTypes.string.isRequired,
  variant: PropTypes.string.isRequired,
};

export default Badge;
