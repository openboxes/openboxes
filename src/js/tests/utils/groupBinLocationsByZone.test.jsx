import { getBinLocationToDisplay, groupBinLocationsByZone } from 'utils/groupBinLocationsByZone';

const mockTranslate = jest.fn((key, defaultValue) => defaultValue);

describe('getBinLocationToDisplay', () => {
  const mockBin = {
    zoneName: 'ZoneA',
    name: 'Bin123',
  };

  it('prepends zoneName to bin name when not already prepended', () => {
    const result = getBinLocationToDisplay(mockBin);
    expect(result).toBe('ZoneA: Bin123');
  });

  it('does not prepend zoneName when already included in name', () => {
    const props = {
      ...mockBin,
      name: 'ZoneA: Bin123',
    };
    const result = getBinLocationToDisplay(props);
    expect(result).toBe('ZoneA: Bin123');
  });

  it('handles missing zoneName safely', () => {
    const props = {
      zoneName: undefined,
      name: 'Bin123',
    };
    const result = getBinLocationToDisplay(props);
    expect(result).toBe('Bin123');
  });

  it('handles missing bin safely', () => {
    const result = getBinLocationToDisplay(undefined);
    expect(result).toBeUndefined();
  });
});

describe('groupBinLocationsByZone', () => {
  const mockBinLocations = [
    {
      id: '1',
      zoneId: 'zone1',
      zoneName: 'ZoneA',
      name: 'Bin1',
    },
    {
      id: '2',
      zoneId: 'zone1',
      zoneName: 'ZoneA',
      name: 'Bin2',
    },
    {
      id: '3',
      zoneId: 'zone2',
      zoneName: 'ZoneB',
      name: 'Bin3',
    },
    {
      id: '4',
      zoneId: null,
      zoneName: null,
      name: 'Bin4',
    },
  ];

  it('groups bins by zoneId and formats options correctly', () => {
    const result = groupBinLocationsByZone(mockBinLocations, mockTranslate);
    expect(result).toHaveLength(3);
    expect(result[0]).toEqual({
      id: 'zone-no-zone',
      name: 'No Zone',
      label: expect.any(Object),
      isDisabled: true,
      options: [{
        id: '4',
        name: 'Bin4',
        label: 'Bin4',
        value: '4',
      }],
    });
    expect(result[1])
      .toEqual({
        id: 'zone-zone1',
        name: 'ZoneA',
        label: expect.any(Object),
        isDisabled: true,
        options: [
          {
            id: '1',
            name: 'ZoneA: Bin1',
            label: 'ZoneA: Bin1',
            value: '1',
          },
          {
            id: '2',
            name: 'ZoneA: Bin2',
            label: 'ZoneA: Bin2',
            value: '2',
          },
        ],
      });
    expect(result[2])
      .toEqual({
        id: 'zone-zone2',
        name: 'ZoneB',
        label: expect.any(Object),
        isDisabled: true,
        options: [{
          id: '3',
          name: 'ZoneB: Bin3',
          label: 'ZoneB: Bin3',
          value: '3',
        }],
      });
  });

  it('handles empty binLocations array safely', () => {
    const result = groupBinLocationsByZone([], mockTranslate);
    expect(result).toEqual([]);
  });

  it('sorts zones alphabetically with No Zone at the start', () => {
    const result = groupBinLocationsByZone(mockBinLocations, mockTranslate);
    expect(result.map((zone) => zone.name)).toEqual(['No Zone', 'ZoneA', 'ZoneB']);
  });

  it('handles missing zoneId and zoneName safely', () => {
    const props = [{
      id: '5',
      zoneId: undefined,
      zoneName: undefined,
      name: 'Bin5',
    }];
    const result = groupBinLocationsByZone(props, mockTranslate);
    expect(result)
      .toEqual([
        {
          id: 'zone-no-zone',
          name: 'No Zone',
          label: expect.any(Object),
          isDisabled: true,
          options: [{
            id: '5',
            name: 'Bin5',
            label: 'Bin5',
            value: '5',
          }],
        },
      ]);
    expect(mockTranslate).toHaveBeenCalledWith('react.cycleCount.noZone', 'No Zone');
  });
});
