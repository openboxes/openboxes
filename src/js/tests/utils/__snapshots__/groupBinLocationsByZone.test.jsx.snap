// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBinLocationToDisplay should match snapshot for sample bin 1`] = `"ZoneA: Bin123"`;

exports[`groupBinLocationsByZone should match snapshot for sample bin locations 1`] = `
[
  {
    "id": "zone-no-zone",
    "isDisabled": true,
    "label": <span
      className="zone-label"
    >
      No Zone
    </span>,
    "name": "No Zone",
    "options": [
      {
        "id": "4",
        "label": "Bin4",
        "name": "Bin4",
        "value": "4",
      },
    ],
  },
  {
    "id": "zone-zone1",
    "isDisabled": true,
    "label": <span
      className="zone-label"
    >
      ZoneA
    </span>,
    "name": "ZoneA",
    "options": [
      {
        "id": "1",
        "label": "ZoneA: Bin1",
        "name": "ZoneA: Bin1",
        "value": "1",
      },
      {
        "id": "2",
        "label": "ZoneA: Bin2",
        "name": "ZoneA: Bin2",
        "value": "2",
      },
    ],
  },
  {
    "id": "zone-zone2",
    "isDisabled": true,
    "label": <span
      className="zone-label"
    >
      ZoneB
    </span>,
    "name": "ZoneB",
    "options": [
      {
        "id": "3",
        "label": "ZoneB: Bin3",
        "name": "ZoneB: Bin3",
        "value": "3",
      },
    ],
  },
]
`;
